File: server/src/inference/model-attribution.mjs
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { isDeepStrictEqual } from "node:util";

export const MODEL_COMPONENTS = Object.freeze([
  "asr",
  "forced-aligner",
  "quran-aligner",
  "acoustic-scorer",
  "calibrator",
]);

export const MODEL_ANALYSIS_BASES = Object.freeze([
  "acoustic",
  "quran-constrained",
  "text-rule",
]);

const SHA256_DIGEST_PATTERN = /^sha256:[a-f0-9]{64}$/;
const COMPONENT_SET = new Set(MODEL_COMPONENTS);
const ANALYSIS_BASIS_SET = new Set(MODEL_ANALYSIS_BASES);

const fail = (message) => {
  throw new Error(`invalid model attribution: ${message}`);
};

const nonEmptyString = (value) => typeof value === "string" && value.length > 0;

/**
 * Runtime validator shared by the server-owned inference and API boundaries.
 * It never fills a default or rewrites a producer identity.
 */
export function validateModelAttribution(value, options = {}) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    fail("value must be an object");
  }
  if (value.schemaVersion !== 1) fail("schemaVersion must be 1");
  if (!COMPONENT_SET.has(value.primaryComponent)) {
    fail(`unknown model component: ${String(value.primaryComponent)}`);
  }
  if (!Array.isArray(value.components) || value.components.length === 0) {
    fail("components must be a non-empty array");
  }

  const seen = new Set();
  const active = new Map();
  for (const record of value.components) {
    if (record === null || typeof record !== "object" || Array.isArray(record)) {
      fail("each component record must be an object");
    }
    if (!COMPONENT_SET.has(record.component)) {
      fail(`unknown model component: ${String(record.component)}`);
    }
    if (seen.has(record.component)) fail(`duplicate model component: ${record.component}`);
    seen.add(record.component);

    if (record.status === "unavailable") {
      if (!nonEmptyString(record.reason)) {
        fail(`unavailable component ${record.component} requires a reason`);
      }
      if (Object.hasOwn(record, "artifactDigest")) {
        fail(`unavailable component ${record.component} cannot claim an artifact digest`);
      }
      continue;
    }
    if (record.status !== "active") {
      fail(`component ${record.component} has an unknown status`);
    }
    if (!nonEmptyString(record.implementationId)) {
      fail(`component ${record.component} requires implementationId`);
    }
    if (!nonEmptyString(record.artifactDigest) || !SHA256_DIGEST_PATTERN.test(record.artifactDigest)) {
      fail(
        `component ${record.component} artifactDigest must be sha256 plus 64 lowercase hex characters`,
      );
    }
    if (!nonEmptyString(record.datasetVersion)) {
      fail(`component ${record.component} requires datasetVersion`);
    }
    if (!ANALYSIS_BASIS_SET.has(record.analysisBasis)) {
      fail(`component ${record.component} has an unknown analysisBasis`);
    }
    if (
      record.calibratorId !== null &&
      (typeof record.calibratorId !== "string" || record.calibratorId.length === 0)
    ) {
      fail(`component ${record.component} has an invalid calibratorId`);
    }
    active.set(record.component, record);
  }

  const primary = active.get(value.primaryComponent);
  if (primary === undefined) fail(`primary component ${value.primaryComponent} must be active`);

  for (const record of active.values()) {
    if (record.calibratorId === null) continue;
    const calibrator = active.get("calibrator");
    if (calibrator === undefined || calibrator.implementationId !== record.calibratorId) {
      fail(`component ${record.component} names an unavailable or mismatched calibrator`);
    }
  }

  if (
    options.legacyModelVersion !== undefined &&
    options.legacyModelVersion !== primary.implementationId
  ) {
    fail(`modelVersion must equal the primary component implementationId ${primary.implementationId}`);
  }

  if (options.expectedDigests !== undefined) {
    for (const [component, expectedDigest] of Object.entries(options.expectedDigests)) {
      if (!COMPONENT_SET.has(component)) fail(`unknown expected model component: ${component}`);
      if (active.get(component)?.artifactDigest !== expectedDigest) {
        fail(`artifact digest mismatch for ${component}`);
      }
    }
  }

  return value;
}

const alignmentBytes = readFileSync(new URL("./alignment.mjs", import.meta.url));
const quranProvenance = JSON.parse(
  readFileSync(
    new URL(
      "../../../packages/quran-data/src/data/full-quran/provenance-v2.json",
      import.meta.url,
    ),
    "utf8",
  ),
);

export const QURAN_ALIGNER_COMPONENT = Object.freeze({
  component: "quran-aligner",
  status: "active",
  implementationId: "quran-constrained-levenshtein@1",
  artifactDigest: `sha256:${createHash("sha256").update(alignmentBytes).digest("hex")}`,
  datasetVersion: `${quranProvenance.canonicalSourceId}:${quranProvenance.edition.identifier}:${quranProvenance.importVersion}`,
  analysisBasis: "quran-constrained",
  calibratorId: null,
});

export const QURAN_ALIGNMENT_ATTRIBUTION = Object.freeze({
  schemaVersion: 1,
  primaryComponent: "quran-aligner",
  components: Object.freeze([QURAN_ALIGNER_COMPONENT]),
});

validateModelAttribution(QURAN_ALIGNMENT_ATTRIBUTION, {
  legacyModelVersion: QURAN_ALIGNER_COMPONENT.implementationId,
});

/**
 * Compose attribution emitted by several bounded inference calls without allowing a later window
 * to silently replace an earlier producer. Repeated component records must be structurally exact;
 * output order follows the closed component vocabulary rather than request timing.
 */
export function mergeModelAttributions(attributions, primaryComponent) {
  if (!Array.isArray(attributions) || attributions.length === 0) {
    fail("model attribution composition requires a non-empty attribution list");
  }
  if (!COMPONENT_SET.has(primaryComponent)) {
    fail(`unknown model component: ${String(primaryComponent)}`);
  }

  const records = new Map();
  for (const attribution of attributions) {
    validateModelAttribution(attribution);
    for (const record of attribution.components) {
      const existing = records.get(record.component);
      if (existing !== undefined && !isDeepStrictEqual(existing, record)) {
        fail(`conflicting model component: ${record.component}`);
      }
      if (existing === undefined) records.set(record.component, { ...record });
    }
  }

  const merged = {
    schemaVersion: 1,
    primaryComponent,
    components: MODEL_COMPONENTS.filter((component) => records.has(component)).map(
      (component) => records.get(component),
    ),
  };
  validateModelAttribution(merged);
  return merged;
}

export function quranAlignmentAttribution(upstreamAttribution = null) {
  if (upstreamAttribution === null) return QURAN_ALIGNMENT_ATTRIBUTION;
  const attribution = mergeModelAttributions(
    [upstreamAttribution, QURAN_ALIGNMENT_ATTRIBUTION],
    "quran-aligner",
  );
  validateModelAttribution(attribution, {
    legacyModelVersion: QURAN_ALIGNER_COMPONENT.implementationId,
  });
  return attribution;
}

File: services/asr-inference/model_attribution.py
"""Runtime-neutral model attribution for the Python inference worker.

This module intentionally imports no torch/model packages, so attribution can be tested and
validated without downloading or loading a checkpoint. It identifies only artifacts the process
can resolve honestly; an alias with no digest is refused instead of being relabelled as evidence.
"""

from __future__ import annotations

import re
from typing import Any, Mapping


MODEL_COMPONENTS = (
    "asr",
    "forced-aligner",
    "quran-aligner",
    "acoustic-scorer",
    "calibrator",
)
MODEL_ANALYSIS_BASES = ("acoustic", "quran-constrained", "text-rule")
_SHA256 = re.compile(r"^sha256:[a-f0-9]{64}$")
_FULL_COMMIT = re.compile(r"^[a-f0-9]{40}$")


def _fail(message: str) -> None:
    raise ValueError(f"invalid model attribution: {message}")


def _require_digest(value: str | None, setting: str) -> str:
    if value is None or not _SHA256.fullmatch(value):
        raise ValueError(f"{setting} must be sha256 plus 64 lowercase hex characters")
    return value


def require_immutable_hf_revision(value: str | None) -> str:
    if value is None or not _FULL_COMMIT.fullmatch(value):
        raise ValueError(
            "ASR_MODEL_REVISION must be a full 40-character lowercase commit hash"
        )
    return value


def _active(
    component: str,
    implementation_id: str,
    artifact_digest: str,
    dataset_version: str,
    analysis_basis: str,
    calibrator_id: str | None = None,
) -> dict[str, Any]:
    return {
        "component": component,
        "status": "active",
        "implementationId": implementation_id,
        "artifactDigest": artifact_digest,
        "datasetVersion": dataset_version,
        "analysisBasis": analysis_basis,
        "calibratorId": calibrator_id,
    }


def _envelope(primary: str, components: list[dict[str, Any]]) -> dict[str, Any]:
    attribution = {
        "schemaVersion": 1,
        "primaryComponent": primary,
        "components": components,
    }
    validate_model_attribution(attribution)
    return attribution


def _digest_from_checkpoint_url(url: str | None) -> str | None:
    if not url:
        return None
    for segment in url.split("/"):
        if re.fullmatch(r"[a-f0-9]{64}", segment):
            return f"sha256:{segment}"
    return None


def build_asr_attribution(
    *,
    model_id: str,
    model_urls: Mapping[str, str],
    package_version: str,
    declared_digest: str | None = None,
    model_revision: str | None = None,
    dataset_version: str = "upstream-training-data-undisclosed",
) -> dict[str, Any]:
    if "/" in model_id:
        artifact_digest = _require_digest(declared_digest, "ASR_MODEL_DIGEST")
        revision = require_immutable_hf_revision(model_revision)
        implementation_id = f"huggingface-transformers-pipeline:{model_id}@{revision}"
    else:
        artifact_digest = _require_digest(
            _digest_from_checkpoint_url(model_urls.get(model_id)),
            f"verified openai-whisper checkpoint URL for {model_id}",
        )
        implementation_id = f"openai-whisper:{model_id}@{package_version}"

    return _envelope(
        "asr",
        [
            _active(
                "asr",
                implementation_id,
                artifact_digest,
                dataset_version,
                "acoustic",
            )
        ],
    )


def build_forced_aligner_attribution(
    model_id: str,
    *,
    declared_digest: str | None = None,
    dataset_version: str = "upstream-training-data-undisclosed",
) -> dict[str, Any]:
    artifact_digest = _require_digest(declared_digest, "FORCE_ALIGN_MODEL_DIGEST")
    return _envelope(
        "forced-aligner",
        [
            _active(
                "forced-aligner",
                f"huggingface-ctc:{model_id}",
                artifact_digest,
                dataset_version,
                "acoustic",
            )
        ],
    )


def build_acoustic_attribution(candidate: Mapping[str, Any]) -> dict[str, Any]:
    model = candidate.get("model")
    if not isinstance(model, Mapping):
        raise ValueError("acoustic candidate must contain model attribution")
    repository = model.get("repository")
    revision = model.get("revision")
    dataset_version = model.get("trainingDataset")
    if not isinstance(repository, str) or not repository:
        raise ValueError("acoustic model repository must be declared")
    if not isinstance(revision, str) or not _FULL_COMMIT.fullmatch(revision):
        raise ValueError("acoustic model revision must be a full commit hash")
    if not isinstance(dataset_version, str) or not dataset_version:
        raise ValueError("acoustic training dataset identity must be declared")
    component = _active(
        "acoustic-scorer",
        f"quran-muaalem:{repository}@{revision}",
        _require_digest(model.get("artifactSha256"), "acoustic artifact digest"),
        dataset_version,
        "acoustic",
    )
    unavailable_calibrator = {
        "component": "calibrator",
        "status": "unavailable",
        "reason": "no held-out calibration artifact has been approved",
    }
    return _envelope("acoustic-scorer", [component, unavailable_calibrator])


def validate_model_attribution(
    value: Any,
    *,
    expected_digests: Mapping[str, str] | None = None,
    legacy_model_version: str | None = None,
) -> dict[str, Any]:
    if not isinstance(value, dict):
        _fail("value must be an object")
    if value.get("schemaVersion") != 1:
        _fail("schemaVersion must be 1")
    primary = value.get("primaryComponent")
    if primary not in MODEL_COMPONENTS:
        _fail(f"unknown model component: {primary}")
    components = value.get("components")
    if not isinstance(components, list) or not components:
        _fail("components must be a non-empty array")

    seen: set[str] = set()
    active: dict[str, dict[str, Any]] = {}
    for record in components:
        if not isinstance(record, dict):
            _fail("each component record must be an object")
        component = record.get("component")
        if component not in MODEL_COMPONENTS:
            _fail(f"unknown model component: {component}")
        if component in seen:
            _fail(f"duplicate model component: {component}")
        seen.add(component)

        if record.get("status") == "unavailable":
            if not isinstance(record.get("reason"), str) or not record["reason"]:
                _fail(f"unavailable component {component} requires a reason")
            if "artifactDigest" in record:
                _fail(f"unavailable component {component} cannot claim an artifact digest")
            continue
        if record.get("status") != "active":
            _fail(f"component {component} has an unknown status")
        if not isinstance(record.get("implementationId"), str) or not record["implementationId"]:
            _fail(f"component {component} requires implementationId")
        if not isinstance(record.get("artifactDigest"), str) or not _SHA256.fullmatch(
            record["artifactDigest"]
        ):
            _fail(f"component {component} has an invalid artifactDigest")
        if not isinstance(record.get("datasetVersion"), str) or not record["datasetVersion"]:
            _fail(f"component {component} requires datasetVersion")
        if record.get("analysisBasis") not in MODEL_ANALYSIS_BASES:
            _fail(f"component {component} has an unknown analysisBasis")
        calibrator_id = record.get("calibratorId")
        if calibrator_id is not None and (not isinstance(calibrator_id, str) or not calibrator_id):
            _fail(f"component {component} has an invalid calibratorId")
        active[component] = record

    primary_record = active.get(primary)
    if primary_record is None:
        _fail(f"primary component {primary} must be active")

    for record in active.values():
        calibrator_id = record["calibratorId"]
        if calibrator_id is None:
            continue
        calibrator = active.get("calibrator")
        if calibrator is None or calibrator["implementationId"] != calibrator_id:
            _fail(f"component {record['component']} names an unavailable or mismatched calibrator")

    if legacy_model_version is not None and legacy_model_version != primary_record["implementationId"]:
        _fail("modelVersion must equal the primary component implementationId")

    if expected_digests is not None:
        for component, digest in expected_digests.items():
            if component not in MODEL_COMPONENTS:
                _fail(f"unknown expected model component: {component}")
            if active.get(component, {}).get("artifactDigest") != digest:
                _fail(f"artifact digest mismatch for {component}")

    return value
